File: LinearSystem/src/Seidel.cpp
#include "Seidel.h"
#include <cmath>
#include <iostream>

vector<double> Seidel::solve(int maxIter, double tol)
{
    int n = rows;

    // if(!isDiagonallyDominant())
    // {
    //     cout<<"Matrix not diagonally dominant ";
    //     makeDiagonallyDominant();
    // }

    vector<double> x(n,0);

    for(int iter=0; iter<maxIter; iter++)
    {
        vector<double> old = x;

        for(int i=0;i<n;i++)
        {
            if((*this)(i,i)==0)
                throw "Zero diagonal in Seidel";

            double sum = (*this)(i,n);

            for(int j=0;j<n;j++)
                if(i!=j)
                    sum -= (*this)(i,j)*x[j];

            x[i] = sum / (*this)(i,i);
        }

        double err = 0;
        for(int i=0;i<n;i++)
            err += fabs(x[i]-old[i]);

        if(err < tol)
            return x;
    }

    return x;
}

File: LinearSystem in Python/src/Seidel.py
import math
from Iteration import Iteration


class Seidel(Iteration):
    def __init__(self, n):
        super().__init__(n)

    def solve(self, max_iter=100, tol=1e-6):
        n = self.rows
        x = [0.0] * n

        for iteration in range(max_iter):
            old = x[:]

            for i in range(n):
                if self.a[i][i] == 0:
                    raise Exception("Zero diagonal in Seidel")

                s = self.a[i][n]
                for j in range(n):
                    if i != j:
                        s -= self.a[i][j] * x[j]
                x[i] = s / self.a[i][i]

            err = 0.0
            for i in range(n):
                err += abs(x[i] - old[i])

            if err < tol:
                return x[:]

        return x
